### PR TITLE
Fix for issue #5086

### PR DESCRIPTION
**Pull request by bummzack**
Improve `Member_Validator` to:
- properly check for existing members.
- allow extensions.
- remove old code and replace with new syntax and add config API.

Fix issue in Group code where Member_Validator was instantiated via "new" which didn't allow injector overrides.
Added unit-tests.

### CODE REVIEW

**Comment by bummzack** [comment]
@dhensby Ok, this one should be good if CI passes.

File: security/Member.php [hunk 1]
```diff
@@ -1842,31 +1860,35 @@ public function __construct() {
...
+		if(!$id && ($ctrl = $this->form->getController())){
+			// get the record when within GridField (Member editing page in CMS)
+			if($ctrl instanceof GridFieldDetailForm_ItemRequest && $record = $ctrl->getRecord()){
+				$id = $record->ID;
+				// set the ID to the data array, so that extensions can also use it
+				$data['ID'] = $id;
+			}
+		}
+
+		$members = Member::get()->filter($identifierField, $data[$identifierField]);
```
Comment (dhensby) [inline security/Member.php]: What if `$data[$identifierField]` is not set?

**Comment by dhensby** [comment]
see https://github.com/silverstripe/silverstripe-framework/pull/5088

File: security/Member.php (hunk 1 above)
Comment (bummzack) [inline security/Member.php]: You mean, if somebody removes the identifier field from the list of required fields? Seems unlikely, but better be safe than sorry… should I create a new PR?

**Comment by bummzack** [comment]
Is there something I missed from the discussion in #5088?

**Comment by dhensby** [comment]
No, in just cross linking so that people can see the discuss ons

File: security/Member.php (hunk 1 above)
Comment (dhensby) [inline security/Member.php]: No. You can just amend this PR and force push. The updates will come into the PR automatically

**Comment by bummzack** [comment]
@dhensby Thanks. I forgot about amending commits. I updated the commit with a safeguard check for `$data[$identifierField]`. I also went ahead and implemented the behavior I proposed in #5086 to solve cases where the validator might not have an `ID` passed with `$data`, yet still "belongs" to a Member.

**Comment by bummzack** [comment]
Any idea what this test failure means: 'Failed asserting that 'JoeEdited' is not equal to <string:JoeEdited>.' ?

**Comment by dhensby** [comment]
I've restarted those tests, bizarre that a couple have it and a couple don't not really sure.

The behat test is more worrying. If you check this out locally and test, does it work in the CMS when you add/save/edit/delete members?

**Comment by bummzack** [comment]
Yeah, everything works fine in the CMS. The warning dialogue now shows up underneath the E-Mail field though… maybe that's causing a problem there?

**Comment by dhensby** [comment]
hmm - I don't think that'd be a problem

**Comment by bummzack** [comment]
Strangely enough all tests have passed with my previous PR… not sure why it failed now. When testing in the CMS, everything works/validates as expected.

**Comment by dhensby** [comment]
someone with more knowledge of PHPUnit will have to explain this one :/

**Comment by bummzack** [comment]
@dhensby I figured out the problem. It is important that the validator checks the `ID` passed via `$data` first. Only if there's no ID passed via `$data` or via controller->record, the member assigned to the Validator should come into play.

**Comment by dhensby** [comment]
Thanks - I'm pretty happy with this and it could lay the foundation for skipping validation on default admins too

**Comment by dhensby** [comment]
@tractorcow thoughts?

**Comment by dhensby** [comment]
oh - as it's a new feature, maybe it'll have to go against 3 :/

**Comment by bummzack** [comment]
You mean, it's a feature that the `Member_Validator` is suddenly working? For me this is more of a Bugfix :)

**Comment by tractorcow** [comment]
Yeah sorry, I meant to say this on the other thread: Can you raise on the 3 branch since it introduces new method API?

**Comment by tractorcow** [comment]
Code is pristine by the way.

**Comment by bummzack** [comment]
Ok, PR against 3 branch is here: #5098. Sorry for the many PRs

**Pull request closed without merging**